### ISSUE TITLE
Add support for service-linked roles

### ISSUE THREAD

**Issue by sshvetsov** [terraform-providers/terraform-provider-aws#921]
AWS recently introduced "[service-linked roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_terms-and-concepts.html?icmpid=docs_iam_console#iam-term-service-linked-role)" - a more secure way of delegating permissions to other AWS services. Trust policy of such roles cannot be modified and only the linked service can assume such role.

### Use Case
These roles are especially useful in AWS accounts where "powerful" services like CloudFormation, Elastic Beanstalk, etc. are used, but IAM admin wants to reduce chances of users/instances assuming traditional service roles.

### Implementation
I'm not sure whether this an enhancement to existing `aws_iam_role` resource or best be treated as a new resource. While output of the API is the same [Role](https://docs.aws.amazon.com/IAM/latest/APIReference/API_Role.html) object, API input parameters are different. It might be more straightforward for end users to create/import and configure this as a separate resource, especially since it cannot be attached to instance profile or assumed by principals other than AWS service.

### References
- [Introducing an Easier Way to Delegate Permissions to AWS Services: Service-Linked Roles](https://aws.amazon.com/blogs/security/introducing-an-easier-way-to-delegate-permissions-to-aws-services-service-linked-roles/)
- [API CreateServiceLinkedRole](https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateServiceLinkedRole.html)
- [aws iam create-service-linked-role](https://docs.aws.amazon.com/cli/latest/reference/iam/create-service-linked-role.html)

**Comment by tv-17** [comment]
Hi,

Any update on this?

**Comment by apparentlymart** [comment]
Given the very different interface for `CreateServiceLinkedRole` compared to the general `CreateRole`, I think a new resource makes sense here. Ideally it'd export the attributes of the resulting role object using the same attribute names and types as `aws_role` uses for input, so that interpolations for attributes of both will be as similar as possible.

```hcl
resource "aws_iam_service_linked_role" "example" {
  aws_service_name = "elasticbeanstalk.amazonaws.com"
}

resource "aws_iam_role_policy" "example" {
  name = "example"
  role = "${aws_iam_service_linked_role.example.name}"

  policy = ...
}

output "role_arn" {
  value = "${aws_iam_service_linked_role.example.arn}"
}
```

The Terraform team at HashiCorp is not currently working on this particular issue. It's likely that we'll get to it at some point since it's a good general UX improvement for using a lot of AWS services -- how to write working assume role policies is a very common question! However, if someone else has the time and motivation to work on this sooner we'd be happy to review a PR for it.

**Comment by jammycakes** [comment]
It should be noted that service-linked roles are automatically created. Also, in at least some cases (certainly, for `aws_appautoscaling_target`), AWS also insists on overriding your choice of role and using the service-linked role instead. This has meant that newly created `aws_appautoscaling_target` resources are appearing as tainted every time.

**Comment by bashoKa** [comment]
Cloudwatch events are affected as well:
does not create AWSServiceRoleForCloudWatchEvents.

**Comment by toddbluhm** [comment]
I just ran into this issue trying to setup a new ecs service. I cannot assign a regular role to a new ecs service anymore, it requires a service linked role.

**Comment by ungureanuvladvictor** [comment]
+1 to this, would be something really useful.

**Comment by bflad** [comment]
The initial `aws_iam_service_linked_role` resource has been merged into master and will release with v1.15.0 of the AWS provider, likely in a week. 🎉

**Comment by bflad** [comment]
This has been released in [version 1.15.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1150-april-18-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
